### DISCUSSION TITLE
Getting error: up.InvalidSelector: Cannot parse selector: .inner1::before

### DISCUSSION THREAD

**Discussion by amitjindal** [unpoly/unpoly#382] [Q&A] [answered]
Hi, I am sorry as I must be doing something wrong but this simple test case is not working.
I am new to unpoly so I am not able to figure out why. I have tried to search forums, bugs, stackoverflow and have not found any such problems so most likely this is my doing.

Below is a simplified testcase. The real use is in a chat application where I am appending received messages to a list of existing messages.

**Code:**

Using CDN:
https://unpkg.com/unpoly@2.6.0/unpoly.min.js
https://unpkg.com/unpoly@2.6.0/unpoly.min.css

```html
<div class="test">Original</div>
<div id="inner1" class="inner1"></div>
```

```js
let htmlString = '<div class="test"></div>';

up.render({
  target: ".inner1",
  content: htmlString
});
 
up.render({
  target: '.inner1:before',
  content: htmlString,
});
```

**Expected:**
This should work and content would be added one after another. However this is not working.
What am I doing wrong?

![ScreenShot 2022-05-13 at 12 19 27 PM](https://user-images.githubusercontent.com/3275983/168229504-1aa63fc4-62ee-4b4e-bcf1-614fca78ffa4.jpg)

**JSFiddle**
https://jsfiddle.net/amitjindal/ceub8d7L/83/

**Comment by triskweline** [comment, marked as answer]
The `:before` and `:after` (single colon) selectors are currently only supported when loading content from a server with `up.render({ url })`, but not with `up.render({ content })`.

This will be fixed in the next version.

Note that your test case could be accomplished with standard browser methods, without the use of a framework like Unpoly:

```
let inner = document.querySelector('.inner')
let htmlString = '<div class="test"></div>'
inner.insertAdjacentHTML('afterbegin', htmlString)
inner.insertAdjacentHTML('beforeend', htmlString)
```

**Reply by amitjindal** [reply]
@triskweline Thank you for your kind reply.

Correct. I can do that with standard javascript. Let me try with code suggested by you.

The root of this problem is a weird behavior I am seeing.

I have a chat application. In that application I am showing user's profile when someone clicks on user's image.

When I use standard javascript to render content the event handler is not getting attached (used up.compiler).
But when I use up.render it does gets attached.

For some reason the DOM is not getting observed or something. No errors on console.

Sample code:
```
    const chatTemplate =  up.element.get("#chat-template").innerHTML;
    let chatContent = up.util.copy(chatTemplate);
    chatContent = chatContent
        .replaceAll("USER_ID", msg.PostedById)
        .replaceAll("USER_IMAGE", msg.avatar)
        .replaceAll("USER_NAME", msg.username)
        .replaceAll("USER_DATE", new Date())
        .replaceAll("USER_MESSAGE", msg.Message);
```

This works (first time only as second gives above mentioned error
```js
    const targetEl = up.element.get("#chat-messages").childElementCount === 0?
        '#chat-messages' : '#chat-messages:after';
    up.render({target: targetEl, content: chatContent});
```

But this does not works.
```js
    const targetEl = up.element.get("#chat-messages");
    targetEl.innerHTML += chatContent;
```

The compiler function that is supposed to display the profile:
```js
up.compiler('[data-ax-toggle="popover"]', function(element) {
   const options = {
      delay: {show: 50, hide: 50},
      html: element.getAttribute('data-bs-html') === "true" ?? false,
      placement: element.getAttribute('data-bs-placement') ?? 'auto'
   };
   const _control = new bootstrap.Popover(element, options);
   return () => _control.dispose();
});
```

**Reply by amitjindal** [reply]
So I tried. With native JS the compiled function (popover) event is not getting attached to inserted node.
Is there any unpoly function I can call to apply those?

I guess, the last option I have is to manually call those functions in the inserted elements.

**Reply by triskweline** [reply]
Content created by `up.render()` gets compiled automatically.

When you create an element outside of Unpoly you can call [`up.hello()`](https://unpoly.com/up.hello) to run compilers.

**Reply by amitjindal** [reply]
@triskweline That helped. Beautiful. Thanks. I do need to figure out how to avoid attaching duplicate event handlers but I can handle that.

I do want to understand a little more about up.compiler as I tried that in a small test case and that works on innerHTML as well.

```html
<div class="inner1">Inner 1</div>
<div class="inner2">Inner 2</div>
```

```js
up.compiler('.test', function(element){
  console.log('Test Works');
});

let htmlString = '<div class="test">Test Div Added</div>';

let inner = document.querySelector('.inner2');
inner.innerHTML = htmlString;

up.render({
  target: ".inner1",
  content: htmlString
});
```

This outputs:

```
Call `up.log.enable()` to enable logging for this session.", "font-family: monospace;color: #777777", "color: #777777"
"Test Works"
"Test Works"
```

So here compiler is indeed working for purely code that ran outside Unpoly.
How do I judge where to use up.hello() and where not to use?

Here's the jsfiddle for this: https://jsfiddle.net/amitjindal/ceub8d7L/97/
